website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,152 @@
+
+
+
+![RaptorX Architecture](/img/blog/2021-02-04-raptorx/raptorx-arch.png)
+
+
+## Metastore Versioned Cache
+
+A Presto coordinator caches table metadata (schema, partition list, and partition info) to avoid long `getPartitions` calls to Hive Metastore. However, Hive table metadata is mutable. Versioning is needed to determine if the cached metadata is valid or not. To achieve that, the coordinator attaches a version number to each cache key-value pair. When a read request comes, the coordinator asks the Hive Metastore to get the partition info (if it's not cached at all) or confirm with Hive Metastore to check if cached info is up to date. Though the roundtrip to Hive Metastore cannot be avoided, the version matching is relatively cheap compared with fetching the entire partition info.
+
+
+
+## File List cache
+
+A Presto coordinator caches file lists in memory to avoid long `listFile` calls to remote storage. This can only be applied to sealed directories. For open partitions, Presto will skip caching those directories to guarantee data freshness. One major use case for open partitions is to support the need of near-realtime ingestion and serving. In such cases, the ingestion engines (e.g., micro batch) will keep writing new files to the open partitions so that Presto can read near-realtime data. Further details like compaction, metastore update, or replication for near-realtime ingestion will be out of the scope of this note.
+
+
+
+## Fragment Result Cache
+
+A Presto worker that is running a leaf stage can decide if to cache the partially computed results on local SSD. This is to prevent duplicated computation upon multiple queries. The most typical use case is to cache the plan fragments on leaf stage with one level of scan, filter, project, or/and aggregation.
+
+For example, suppose a user sends the following query, where `ds` is a partition column:
+
+```sql
+select sum(col) from T where ds between '2021-01-01' and '2021-01-03'
+```
+
+The partially computed sum for each of `2021-01-01`, `2021-01-02`, and `2021-01-03` partitions (or more precisely the corresponding files) will be cached on leaf workers forming a "fragment result". Suppose the user sends another query:
+
+```sql
+select sum(col) from T where ds between '2021-01-01' and '2021-01-05'
+```
+
+Then, the leaf worker will directly fetch the fragment result for `2021-01-01`, `2021-01-02`, and `2021-01-03` from cache and just compute the partial sum for `2021-01-04` and `2021-01-05`.
+
+Note that the fragment result is based on the leaf query fragment, which could be highly flexible as users can add or remove filters or projects. The above example shows we can easily handle filters with only partition columns. In order to avoid the cache miss caused by frequently changing non-partition column filters, we introduced partition statistics based pruning. Consider the following query, where time is a non-partition column:
```

Comment:
```suggestion
Note that the fragment result is based on the leaf query fragment, which could be highly flexible as users can add or remove filters or projections. The above example shows we can easily handle filters with only partition columns. In order to avoid the cache miss caused by frequently changing non-partition column filters, we introduced partition statistics based pruning. Consider the following query, where time is a non-partition column:
```